DPMMA-2550 Point Groups API

patrykgruszka opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 5.x)
* a.b for any bug fixes (e.g. 4.4, 5.1)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ N ]
| New feature/enhancement? (use the a.x branch)      | [ Y ]
| Deprecations?                          | [ N ]
| BC breaks? (use the c.x branch)        | [ N ]
| Automated tests included?              | [ Y ] <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | https://github.com/mautic/developer-documentation-new/pull/185
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

#### Description:
This PR adds API endpoints for Point Groups management.
```
mautic_api_pointGroups_getall          GET     /api/points/groups                                                                               
mautic_api_pointGroups_getone          GET     /api/points/groups/{id}                                                                          
mautic_api_pointGroups_new             POST    /api/points/groups/new                                                                                                                                        
mautic_api_pointGroups_editput         PUT     /api/points/groups/{id}/edit                                                                     
mautic_api_pointGroups_editpatch       PATCH   /api/points/groups/{id}/edit                                                                                                                           
mautic_api_pointGroups_delete          DELETE  /api/points/groups/{id}/delete                                                                   
mautic_api_getcontactpointgroups       GET     /api/contacts/{contactId}/points/groups                                                          
mautic_api_getcontactpointgroup        GET     /api/contacts/{contactId}/points/groups/{groupId}                                                
mautic_api_adjustcontactgrouppoints    POST    /api/contacts/{contactId}/points/groups/{groupId}/{operator}/{value} 
```

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to test this PR:

<!--
This part is really important. If you want your PR to be merged, take the time to write very clear, annotated and step by step test instructions. Do not assume any previous knowledge - testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))

2. **Create a new point group**:
   - **Endpoint**: `POST https://your-api-host.com/api/points/groups/new`
   - **Payload**: JSON payload with the following fields:
     ```json
     {
         "name": "New Point Group",
         "description": "Description of the new point group"
     }
     ```

3. **Retrieve all point groups**:
   - **Endpoint**: `GET https://your-api-host.com/api/points/groups`

4. **Update the created point group**:
   - **Endpoint**: `PATCH https://your-api-host.com/api/points/groups/{id}/edit`
   - **Payload**: JSON payload with the fields to update:
     ```json
     {
         "name": "Updated Point Group Name",
         "description": "Updated description of the point group"
     }
     ```

5. **Delete the created point group**:
   - **Endpoint**: `DELETE https://your-api-host.com/api/points/groups/{id}/delete`

6. **Get contact group points**:
   - **Endpoint**: `GET https://your-api-host.com/api/contacts/{leadId}/points/groups`

7. **Adjust contact group points**:
   - **Plus operation**:
     - **Endpoint**: `POST https://your-api-host.com/api/contacts/{leadId}/points/groups/{groupId}/plus/{value}`
   - **Minus operation**:
     - **Endpoint**: `POST https://your-api-host.com/api/contacts/{leadId}/points/groups/{groupId}/minus/{value}`
   - **Set operation**:
     - **Endpoint**: `POST https://your-api-host.com/api/contacts/{leadId}/points/groups/{groupId}/set/{value}`
   - **Divide operation**:
     - **Endpoint**: `POST https://your-api-host.com/api/contacts/{leadId}/points/groups/{groupId}/divide/{value}`
   - **Multiply operation**:
     - **Endpoint**: `POST https://your-api-host.com/api/contacts/{leadId}/points/groups/{groupId}/times/{value}`

When testing these endpoints, ensure that you replace placeholders such as `{leadId}`, `{groupId}`, and `{value}` with actual IDs and values as required for your specific test cases. Additionally, for endpoints that require a request body, provide the appropriate JSON payload with the necessary fields and values.

patrykgruszka committed:
feat: [DPMMA-2550] point groups api

patrykgruszka committed:
feat: [DPMMA-2550] contact group score api

patrykgruszka committed:
feat: [DPMMA-2550] contact group score api changelog

patrykgruszka committed:
feat: [DPMMA-2550] contact group score api tests

patrykgruszka committed:
feat: [DPMMA-2550] contact single group score api and refactor

patrykgruszka committed:
feat: [DPMMA-2550] change group entity name in api

patrykgruszka committed:
fix: [DPMMA-2550] fix rector for PointGroupsApiController

patrykgruszka committed:
Merge branch '5.x' into DPMMA-2550_point-groups-api

RCheesley committed:
Merge branch '5.x' into DPMMA-2550_point-groups-api

RCheesley reviewed (dismissed):
Tested all endpoints (thank you SO MUCH for the awesome documentation in the PR @patrykgruszka it really helps testers when you provide both the path and the snippet!).

All working as expected. Thanks for the great work! 🚀

escopecz commented on app/bundles/PointBundle/Config/config.php:
```diff
@@ -63,6 +63,25 @@
+                'standard_entity' => true,
+                'name'            => 'pointGroups',
+                'path'            => '/points/groups',
+                'controller'      => \Mautic\PointBundle\Controller\Api\PointGroupsApiController::class,
```
As per the new CS Fixer version:
```suggestion
                'controller'      => Mautic\PointBundle\Controller\Api\PointGroupsApiController::class,
```

escopecz commented on app/bundles/PointBundle/Controller/Api/PointGroupsApiController.php:
```diff
@@ -0,0 +1,150 @@
+     */
+    protected $model;
+
+    /** @phpstan-ignore-next-line */
```
Why is this needed? If it cannot be resolved, could you add the reason to the comment?

patrykgruszka replied:
Sure, I've added a comment. It is caused by the parent class.

escopecz commented on app/bundles/PointBundle/Controller/Api/PointGroupsApiController.php:
```diff
@@ -0,0 +1,150 @@
+ */
+class PointGroupsApiController extends CommonApiController
+{
+    /**
+     * @var PointGroupModel
+     */
+    protected $model;
```
```suggestion
    private PointGroupModel $model;
```
Or is there some weird inheritance thing going on?

patrykgruszka replied:
Yes, we can't define the type here.

escopecz commented on app/bundles/PointBundle/Controller/Api/PointGroupsApiController.php:
```diff
@@ -0,0 +1,150 @@
+<?php
+
```
```suggestion

declare(strict_types=1);

```

escopecz commented on app/bundles/PointBundle/Tests/Controller/Api/PointGroupsApiControllerTest.php:
```diff
@@ -0,0 +1,182 @@
+<?php
+
```
```suggestion

declare(strict_types=1);

```

escopecz commented on app/bundles/PointBundle/Tests/Controller/Api/PointGroupsApiControllerTest.php:
```diff
@@ -0,0 +1,182 @@
+        $this->client->request('DELETE', "/api/points/groups/{$createdData['id']}/delete");
+        $deleteResponse = $this->client->getResponse();
+
+        $this->assertSame(200, $deleteResponse->getStatusCode());
```
```suggestion
        $this->assertSame(Response::HTTP_OK, $deleteResponse->getStatusCode());
```

escopecz reviewed (changes requested):
The code looks great! I have a few small suggestions. The functional tests are testing the happy paths. It would be great to have tests that check also 404s, xss and similar. But happy path tests are still way better than none 👍

patrykgruszka committed:
fix: [DPMMA-2550] point groups api improvements and extra tests

patrykgruszka committed:
Merge branch '5.x' into DPMMA-2550_point-groups-api

patrykgruszka commented:
Thanks! I've updated the code and added some extra tests :)

Would you have some time to look at the related PR in the API library? https://github.com/mautic/api-library/pull/322

escopecz reviewed (approved):
I see no issue in the code. The Point group/score is covered with functional tests 👍

escopecz committed:
Merge branch '5.x' into DPMMA-2550_point-groups-api
